File: backend/models/ai_models.go
package models

type AIGeneratedGraph struct {
	Nodes       []AINode `json:"nodes"`
	Edges       []AIEdge `json:"edges"`
	Explanation string   `json:"explanation"`
	Suggestions []string `json:"suggestions"`
}

type AINode struct {
	ID       string    `json:"id"`
	Type     string    `json:"type"`
	Label    string    `json:"label"`
	Capacity int       `json:"capacity"`
	Latency  int       `json:"latency"`
	Position XPosition `json:"position"`

	// FIX 2: Node-type-specific fields the AI can now set
	HitRatio       float64 `json:"hit_ratio,omitempty"`       // Cache/CDN
	MaxConnections int     `json:"max_connections,omitempty"` // Database/Storage
	MaxQueueDepth  int     `json:"max_queue_depth,omitempty"` // Queue
}

type XPosition struct {
	X float64 `json:"x"`
	Y float64 `json:"y"`
}

type AIEdge struct {
	Source string `json:"source"`
	Target string `json:"target"`
}

File: backend-python/app/models/ai_models.py
"""
Models for AI-generated system designs (Grok/LLM responses).

Go → Python mapping:
  - Go nested struct `XPosition` → Pydantic `XPosition` model
  - Go AINode/AIEdge/AIGeneratedGraph → Pydantic models with Optional fields
"""

from __future__ import annotations

from typing import Optional

from pydantic import BaseModel, Field


class XPosition(BaseModel):
    """2D canvas coordinate for frontend node placement."""

    x: float
    y: float


class AINode(BaseModel):
    """A node as returned by the AI design endpoint."""

    id: str
    type: str
    label: str
    capacity: int = 0
    latency: int = 0
    position: XPosition

    # Node-type-specific fields the AI can set
    hit_ratio: Optional[float] = Field(default=None, ge=0.0, le=1.0)
    max_connections: Optional[int] = Field(default=None, ge=0)
    max_queue_depth: Optional[int] = Field(default=None, ge=0)


class AIEdge(BaseModel):
    """Edge in the AI-generated graph."""

    source: str
    target: str


class AIGeneratedGraph(BaseModel):
    """Complete response from the AI design generation endpoint."""

    nodes: list[AINode]
    edges: list[AIEdge]
    explanation: str = ""
    suggestions: list[str] = Field(default_factory=list)
